README.md
# Goflight
![](https://goreportcard.com/badge/github.com/marcelblijleven/goflight)

Go client for the [OpenSky REST API](http://opensky-network.org).

## Example
Get all states

```go
package main

import (
	"fmt"
	"github.com/marcelblijleven/goflight"
	"net/http"
	"time"
)

func main() {
	httpClient := http.Client{Timeout: time.Second * 30}
	client, _ := goflight.NewClient(
		"user@email.com",
		"tops3cr3t",
		&httpClient,
	)

	timeParam := time.Now()
	icao24 := "3c6444"

	resp, _ := client.States.GetAllStates(timeParam, icao24)
    fmt.Println(resp.States[0].ICAO24)
}
```

## Disclaimer
This project is not affiliated with the Opensky Network


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.0.1 (2020-04-07)

## 1.1.0 (2020-04-07)


### Features

* add functionality for /api/flights/all ([#12](https://github.com/marcelblijleven/goflight/issues/12)) ([0b72013](https://github.com/marcelblijleven/goflight/commit/0b720132c85a4279dc1d56ec6fe93fcf7c03e524))

### 1.0.1 (2020-04-06)


### Bug Fixes

* fixed ineffectual assignment to err variable ([00062d7](https://github.com/marcelblijleven/goflight/commit/00062d71ce6d3d046bd700c7ca22e93a03eaf7ac))


export_test.go
package goflight

// SetBaseURL exports the setBaseURL func on client, but only in tests
var SetBaseURL = (*Client).setBaseURL


states_test.go
package goflight_test

import (
	"encoding/json"
	"errors"
	"github.com/marcelblijleven/goflight"
	"io/ioutil"
	"net/url"
	"testing"
	"time"
)

var stateVectorTests = []struct {
	inputFile string
}{
	{"./mocks/state_vector.json"},
	{"./mocks/state_vector_null_values.json"},
}

func TestStateVector_UnmarshalJSON(t *testing.T) {
	for _, tt := range stateVectorTests {
		t.Run(tt.inputFile, func(t *testing.T) {
			data, err := ioutil.ReadFile(tt.inputFile)

			if err != nil {
				t.Fatal(err.Error())
			}

			var vector goflight.StateVector

			if err = json.Unmarshal(data, &vector); err != nil {
				t.Errorf("unexpected error for %q: %v", tt.inputFile, err.Error())
			}

			// Unmarshal mock data onto slice of interface
			// to check values
			var tmp []interface{}

			if err = json.Unmarshal(data, &tmp); err != nil {
				t.Fatal(err.Error())
			}
		})
	}
}

func TestStateVector_UnmarshalJSON_IncorrectLength(t *testing.T) {
	data, err := ioutil.ReadFile("./mocks/state_vector_incorrect_length.json")

	if err != nil {
		t.Fatal(err.Error())
	}

	var vector goflight.StateVector

	if err = json.Unmarshal(data, &vector); err == nil {
		t.Errorf("expected error to be non nil")
	}
}

var getStatesInputs = []struct {
	icao24Input string
	timeInput   time.Time
}{
	{"", time.Time{}},
	{"c0ffee", time.Date(2020, time.July, 6, 0, 0, 0, 0, time.UTC)},
}

func TestStatesService_GetAllStates(t *testing.T) {
	mockResponseBody, err := ioutil.ReadFile("./mocks/states.json")

	if err != nil {
		t.Fatal("unexpected error in retrieving mock response body")
	}

	mockHandler := CreateTestHandler(200, mockResponseBody)
	mockClient, closeServer := HTTPTestClient(mockHandler)
	defer closeServer()

	for _, tt := range getStatesInputs {
		t.Run(tt.icao24Input, func(t *testing.T) {
			client, err := goflight.NewClient("", "", mockClient)
			u, _ := url.Parse("http://example.com")
			goflight.SetBaseURL(client, u)

			if err != nil {
				t.Fatal("unexpected error in setting up Goflight client")
			}

			response, err := client.States.GetAllStates(tt.timeInput, tt.icao24Input)

			if err != nil {
				t.Error(err.Error())
			}

			var expectedTime int64
			expectedTime = 1586031310

			if response.Time != 1586031310 {
				t.Errorf("expected %v to equal %v", response.Time, expectedTime)
			}

			if len(response.States) != 6 {
				t.Error("expect length of states in response to equal 6")
			}
		})
	}
}

func TestStatesService_GetOwnStates(t *testing.T) {
	mockResponseBody, err := ioutil.ReadFile("./mocks/states.json")

	if err != nil {
		t.Fatal("unexpected error in retrieving mock response body")
	}

	mockHandler := CreateTestHandler(200, mockResponseBody)
	mockClient, closeServer := HTTPTestClient(mockHandler)
	defer closeServer()

	for _, tt := range getStatesInputs {
		t.Run(tt.icao24Input, func(t *testing.T) {
			client, err := goflight.NewClient("user", "password", mockClient)
			u, _ := url.Parse("http://example.com")
			goflight.SetBaseURL(client, u)

			if err != nil {
				t.Fatal("unexpected error in setting up Goflight client")
			}

			response, err := client.States.GetAllStates(tt.timeInput, tt.icao24Input)

			if err != nil {
				t.Error(err.Error())
			}

			var expectedTime int64
			expectedTime = 1586031310

			if response.Time != 1586031310 {
				t.Errorf("expected %v to equal %v", response.Time, expectedTime)
			}

			if len(response.States) != 6 {
				t.Error("expect length of states in response to equal 6")
			}
		})
	}
}

func TestStatesService_GetOwnStates_Authentication(t *testing.T) {
	// Check for unauthorized access
	mockHandler := CreateTestHandler(403, nil)
	mockClient, closeServer := HTTPTestClient(mockHandler)
	defer closeServer()

	client, err := goflight.NewClient("unauthorizedUser", "secret", mockClient)

	if err != nil {
		t.Fatal("unexpected error in setting up Goflight client")
	}

	_, err = client.States.GetOwnStates(time.Time{}, "")

	if !errors.As(err, &goflight.ErrUnauthorizedAccess) {
		t.Errorf("expected error to be: %v", goflight.ErrUnauthorizedAccess.Error())
	}

	// Check for invalid credentials
	client, err = goflight.NewClient("", "", mockClient)

	if err != nil {
		t.Fatal("unexpected error in setting up Goflight client")
	}

	_, err = client.States.GetOwnStates(time.Time{}, "")

	if !errors.Is(err, goflight.ErrInvalidCredentials) {
		t.Errorf("expected error to be: %v", goflight.ErrUnauthorizedAccess.Error())
	}
}


flights_test.go
package goflight_test

import (
	"encoding/json"
	"errors"
	"github.com/marcelblijleven/goflight"
	"io/ioutil"
	"net/url"
	"testing"
	"time"
)

func TestFlight_UnmarshalJSON(t *testing.T) {
	data, err := ioutil.ReadFile("./mocks/flights.json")

	if err != nil {
		t.Fatal("unexpected error while reading mock file flights.json")
	}

	var result []goflight.Flight

	if err = json.Unmarshal(data, &result); err != nil {
		t.Error(err.Error())
	}

	if len(result) != 2 {
		t.Error("expected length of flights to be 2")
	}
}

var getFlightInTimeTests = []struct {
	label       string
	begin       time.Time
	end         time.Time
	errExpected error
	statusCode  int
	mockFile    string
}{
	{
		"Correct call",
		time.Now(),
		time.Now().Add(time.Hour),
		nil,
		200,
		"./mocks/flights.json",
	},
	{
		"Time range too big",
		time.Now(),
		time.Now().Add(time.Hour * 3),
		goflight.ErrTimeRangeTooBig,
		0,
		"",
	},
	{
		"End before begin",
		time.Now().Add(time.Hour),
		time.Now(),
		goflight.ErrEndBeforeBegin,
		0,
		"",
	},
	{
		"No results found",
		time.Now(),
		time.Now().Add(time.Hour),
		nil,
		404,
		"",
	},
	{
		"Server error",
		time.Now(),
		time.Now().Add(time.Hour),
		errors.New("500 Internal Server Error"),
		500,
		"",
	},
}

func TestFlightService_GetFlightsInTime(t *testing.T) {
	for _, tt := range getFlightInTimeTests {
		t.Run(tt.label, func(t *testing.T) {
			if tt.mockFile == "" && tt.errExpected != nil {
				// Checking returns errors
				if tt.statusCode == 0 {
					// Checking errors where calls to the http client are not made
					client, err := goflight.NewClient("", "", nil)

					if err != nil {
						t.Fatal("unexpected error setting up new Goflight client")
					}

					flights, err := client.Flights.GetFlightsInTime(tt.begin, tt.end)

					if err == nil {
						t.Error("expected err to be non nil")
					}

					if !errors.As(err, &tt.errExpected) {
						t.Errorf("expected error to be as %v", tt.errExpected.Error())
					}

					if flights != nil {
						t.Error("expected flights to be nil")
					}
				}

				if tt.statusCode != 0 {
					// Checking errors where calls to the http client are made
					mockHandler := CreateTestHandler(tt.statusCode, nil)
					mockHTTPClient, closeServer := HTTPTestClient(mockHandler)
					defer closeServer()

					client, err := goflight.NewClient("", "", mockHTTPClient)
					u, _ := url.Parse("http://example.com")
					goflight.SetBaseURL(client, u)

					if err != nil {
						t.Fatal("unexpected error while creating new Gofight client")
					}

					flights, err := client.Flights.GetFlightsInTime(tt.begin, tt.end)

					if err == nil {
						t.Error("expected err to be non nil")
					}

					if err.Error() != tt.errExpected.Error() {
						t.Errorf("expected %v to equal %v", tt.errExpected.Error(), err.Error())
					}

					if flights != nil {
						t.Error("expected flights to be nil")
					}
				}
			}

			if tt.mockFile != "" {
				// Checking requests where a mock response is expected
				body, err := ioutil.ReadFile(tt.mockFile)

				if err != nil {
					t.Fatal("unexpected error while reading mock file flights.json")
				}

				mockHandler := CreateTestHandler(tt.statusCode, body)
				mockHTTPClient, closeServer := HTTPTestClient(mockHandler)
				defer closeServer()

				client, err := goflight.NewClient("", "", mockHTTPClient)
				u, _ := url.Parse("http://example.com")
				goflight.SetBaseURL(client, u)

				if err != nil {
					t.Fatal("unexpected error while creating new Gofight client")
				}

				flights, err := client.Flights.GetFlightsInTime(tt.begin, tt.end)

				if err != nil {
					t.Error("expected err to be nil")
				}

				if flights == nil {
					t.Error("expected flights to be non nil")
				}

				if len(flights) != 2 {
					t.Error("expected flights to have a length of 2")
				}
			}
		})
	}
}


flights.go
package goflight

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

const flightsPrefix string = "/api/flights/"

type flightService struct {
	client *Client
}

// Flight represents a flight in an /api/flights/* response
type Flight struct {
	ICAO24                                string  `json:"icao24"`                           // Unique ICAO 24-bit address of the transponder in hex string representation. All letters are lower case.
	FirstSeen                             int64   `json:"firstSeen"`                        // Estimated time of departure for the flight as Unix time (seconds since epoch).
	EstDepartureAirport                   *string `json:"estDepartureAirport"`              // ICAO code of the estimated departure airport. Can be null if the airport could not be identified.
	LastSeen                              int64   `json:"lastSeen"`                         // Estimated time of arrival for the flight as Unix time (seconds since epoch)
	EstArrivalAirport                     *string `json:"estArrivalAirport"`                // ICAO code of the estimated arrival airport. Can be null if the airport could not be identified.
	CallSign                              *string `json:"callsign"`                         // Callsign of the vehicle (8 chars). Can be null if no callsign has been received. If the vehicle transmits multiple callsigns during the flight, we take the one seen most frequently
	EstDepartureAirportHorizontalDistance *int64  `json:"estDepartureAirportHorizDistance"` // Horizontal distance of the last received airborne position to the estimated departure airport in meters
	EstDepartureAirportVerticalDistance   *int64  `json:"estDepartureAirportVertDistance"`  // Vertical distance of the last received airborne position to the estimated departure airport in meters
	EstArrivalAirportHorizontalDistance   *int64  `json:"estArrivalAirportHorizDistance"`   // Horizontal distance of the last received airborne position to the estimated arrival airport in meters
	EstArrivalAirportVerticalDistance     *int64  `json:"estArrivalAirportVertDistance"`    // Vertical distance of the last received airborne position to the estimated arrival airport in meters
	DepartureAirportCandidatesCount       int64   `json:"departureAirportCandidatesCount"`  // Number of other possible departure airports. These are airports in short distance to estDepartureAirport.
	ArrivalAirportCandidatesCount         int64   `json:"arrivalAirportCandidatesCount"`    // Number of other possible departure airports. These are airports in short distance to estArrivalAirport.
}

func (f *flightService) GetFlightsInTime(begin, end time.Time) ([]Flight, error) {
	endpoint, err := url.Parse(flightsPrefix + "all")

	if err != nil {
		return nil, err
	}

	if end.Before(begin) {
		return nil, ErrEndBeforeBegin
	}

	if end.Sub(begin) > time.Hour*2 {
		return nil, ErrTimeRangeTooBig
	}

	u := f.client.baseURL.ResolveReference(endpoint)
	params := url.Values{}
	params.Add("begin", strconv.FormatInt(begin.Unix(), 10))
	params.Add("end", strconv.FormatInt(end.Unix(), 10))
	u.RawQuery = params.Encode()

	req, err := http.NewRequest("GET", u.String(), nil)

	if err != nil {
		return nil, err
	}

	username, okUser := checkString(f.client.username)
	password, okPassword := checkString(f.client.password)

	if okUser && okPassword {
		req.SetBasicAuth(username, password)
	}

	resp, err := f.client.httpClient.Do(req)

	if err != nil {
		return nil, err
	}

	if resp.StatusCode/100 != 2 {
		if resp.StatusCode == 404 {
			return []Flight{}, nil
		}

		return nil, fmt.Errorf(resp.Status)
	}

	data, err := ioutil.ReadAll(resp.Body)

	if err != nil {
		return nil, err
	}

	defer resp.Body.Close()

	var result []Flight

	if err = json.Unmarshal(data, &result); err != nil {
		return nil, err
	}

	return result, nil
}


goflight.go
package goflight

import (
	"net/http"
	"net/url"
	"time"
)

const (
	baseURL = "https://opensky-network.org"
)

// Client represents the API client used to communicate with the Opensky API
type Client struct {
	httpClient *http.Client
	baseURL    *url.URL
	username   string
	password   string

	States  *statesService
	Flights *flightService
}

// NewClient creates a new client with the provided credentials
func NewClient(username, password string, httpClient *http.Client) (*Client, error) {
	if httpClient == nil {
		httpClient = &http.Client{Timeout: time.Second * 30}
	}

	u, err := url.Parse(baseURL)
	if err != nil {
		return nil, err
	}

	c := &Client{
		httpClient: httpClient,
		baseURL:    u,
		username:   username,
		password:   password,
	}

	c.States = &statesService{client: c}
	c.Flights = &flightService{client: c}

	return c, nil
}

// GetBaseURL returns the base url of the client
func (c *Client) GetBaseURL() *url.URL {
	return c.baseURL
}

// GetCredentials returns the username and password of the client
func (c *Client) GetCredentials() (username string, password string) {
	return c.username, c.password
}

func (c *Client) setBaseURL(u *url.URL) {
	c.baseURL = u
}


states.go
package goflight

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

type statesService struct {
	client *Client
}

// StateVector represents the state of the aircraft at a given time
type StateVector struct {
	ICAO24         string   // Unique ICAO 24-bit address of the transponder in hex string representation.
	Callsign       *string  // Callsign of the vehicle (8 chars). Can be nil if no callsign has been received.
	OriginCountry  string   // Country name inferred from the ICAO 24-bit address.
	TimePosition   *int64   // Unix timestamp (seconds) for the last position update. Can be nil if no position report was received by OpenSky within the past 15s.
	LastContact    int64    // Unix timestamp (seconds) for the last update in general. This field is updated for any new, valid message received from the transponder.
	Longitude      *float64 // WGS-84 longitude in decimal degrees. Can be nil.
	Latitude       *float64 // WGS-84 latitude in decimal degrees. Can be nil.
	BaroAltitude   *float64 // Barometric altitude in meters. Can be nil.
	OnGround       bool     // Boolean value which indicates if the position was retrieved from a surface position report.
	Velocity       *float64 // Velocity over ground in m/s. Can be nil.
	TrueTrack      *float64 // True track in decimal degrees clockwise from north (north=0°). Can be nil.
	VerticalRate   *float64 // Vertical rate in m/s. A positive value indicates that the airplane is climbing, a negative value indicates that it descends. Can be nil.
	Sensors        *[]int   // IDs of the receivers which contributed to this state vector. Is nil if no filtering for sensor was used in the request.
	GeoAltitude    *float64 // Geometric altitude in meters. Can be nil.
	Squawk         *string  // The transponder code aka Squawk. Can be nil.
	Spi            *bool    // Whether flight status indicates special purpose indicator. Can be nil
	PositionSource int      // Origin of this state’s position: 0 = ADS-B, 1 = ASTERIX, 2 = MLAT
}

// StatesResponse is the response retrieved from the /api/states/all and /api/states/own endpoints
type StatesResponse struct {
	Time   int64         `json:"time"`
	States []StateVector `json:"states"`
}

// UnmarshalJSON unmarshals the provided []byte onto a StateVector struct
func (s *StateVector) UnmarshalJSON(buf []byte) error {
	tmp := []interface{}{
		&s.ICAO24,
		&s.Callsign,
		&s.OriginCountry,
		&s.TimePosition,
		&s.LastContact,
		&s.Longitude,
		&s.Latitude,
		&s.BaroAltitude,
		&s.OnGround,
		&s.Velocity,
		&s.TrueTrack,
		&s.VerticalRate,
		&s.Sensors,
		&s.GeoAltitude,
		&s.Squawk,
		&s.Spi,
		&s.PositionSource,
	}

	expectedLen := len(tmp)

	if err := json.Unmarshal(buf, &tmp); err != nil {
		return err
	}

	if actual, expected := len(tmp), expectedLen; actual != expected {
		return errors.New("incorrect number of fields in StateVector json")
	}

	return nil
}

func (s *statesService) getStatesRequest(endpoint string, timeParam time.Time, icao24 string) (*http.Request, error) {
	method := "GET"

	e, err := url.Parse(endpoint)

	if err != nil {
		return nil, err
	}

	u := s.client.baseURL.ResolveReference(e)

	req, err := http.NewRequest(method, u.String(), nil)

	if err != nil {
		return nil, err
	}

	if timeParam, ok := checkTime(timeParam); ok {
		req.URL.Query().Add("time", strconv.FormatInt(timeParam.Unix(), 10))
	}

	if icao24, ok := checkString(icao24); ok {
		req.URL.Query().Add("icao24", icao24)
	}

	req.URL.RawQuery = req.URL.Query().Encode()

	return req, nil
}

// GetAllStates returns the response of /api/states/all
func (s *statesService) GetAllStates(time time.Time, icao24 string) (StatesResponse, error) {
	endpoint := "/api/states/all"
	req, err := s.getStatesRequest(endpoint, time, icao24)

	if err != nil {
		return StatesResponse{}, err
	}

	username, okUser := checkString(s.client.username)
	password, okPass := checkString(s.client.password)

	if okUser && okPass {
		// Authentication is not required for this endpoint
		req.SetBasicAuth(username, password)
	}

	resp, err := s.client.httpClient.Do(req)
	var statesResponse StatesResponse

	if err != nil {
		return statesResponse, err
	}

	if resp.StatusCode/100 != 2 {
		if resp.StatusCode == 403 {
			return statesResponse, ErrUnauthorizedAccess
		}

		return statesResponse, fmt.Errorf("%v - %v", resp.StatusCode, resp.Status)
	}

	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)

	if err != nil {
		return statesResponse, err
	}

	if err = json.Unmarshal(body, &statesResponse); err != nil {
		return statesResponse, err
	}

	return statesResponse, nil
}

// GetOwnStates returns the response of /api/states/own
func (s *statesService) GetOwnStates(time time.Time, icao24 string) (*StatesResponse, error) {
	endpoint := "/api/states/own"
	req, err := s.getStatesRequest(endpoint, time, icao24)

	if err != nil {
		return nil, err
	}

	username, okUser := checkString(s.client.username)
	password, okPass := checkString(s.client.password)

	if !(okUser && okPass) {
		// Authentication is required for this endpoint
		return nil, ErrInvalidCredentials
	}

	req.SetBasicAuth(username, password)

	resp, err := s.client.httpClient.Do(req)

	if err != nil {
		return nil, err
	}

	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)

	if err != nil {
		return nil, err
	}

	var response *StatesResponse
	if err = json.Unmarshal(body, response); err != nil {
		return nil, err
	}

	return response, nil
}
